svf-llvm/lib/ObjTypeInference.cpp

```diff
@@ -590,69 +592,77 @@ u32_t ObjTypeInference::objTyToNumFields(const Type *objTy)
  * @param thisPtr
- * @return
+ * @return a set of all possible type names that :param:`thisPtr` could point to
  */
 Set<std::string> &ObjTypeInference::inferThisPtrClsName(const Value *thisPtr)
 {
     auto it = _thisPtrClassNames.find(thisPtr);
     if (it != _thisPtrClassNames.end()) return it->second;
 
     Set<std::string> names;
-    auto insertClassNames = [&names](Set<std::string> &classNames)
+
+    // Lambda for checking a function is a valid name source & extracting a class name from it
+    auto addNamesFromFunc = [&names](const Function *func) -> void
     {
-        names.insert(classNames.begin(), classNames.end());
+        ABORT_IFNOT(isClsNameSource(func), "Func is invalid class name source: " + dumpValueAndDbgInfo(func));
+        for (auto name : extractClsNamesFromFunc(func)) names.insert(name);
     };
 
-    // backward find heap allocations or class name sources
-    Set<const Value *> &vals = bwFindAllocOrClsNameSources(thisPtr);
-    for (const auto &val: vals)
+    // Lambda for getting callee & extracting class name for calls to constructors/destructors/template funcs
+    auto addNamesFromCall = [&names, &addNamesFromFunc](const CallBase *call) -> void
     {
+        ABORT_IFNOT(isClsNameSource(call), "Call is invalid class name source: " + dumpValueAndDbgInfo(call));
+
+        const auto *func = call->getCalledFunction();
+        if (isDynCast(func)) names.insert(extractClsNameFromDynCast(call));
+        else addNamesFromFunc(func);
+    };
+
+    // Walk backwards to find all valid source sites for the pointer (e.g. stack/global/heap variables)
+    for (const auto &val: bwFindAllocOrClsNameSources(thisPtr))
+    {
+        // A source site is either a constructor/destructor/template function from which the class name can be
+        // extracted; a call to a C++ constructor/destructor/template function from which the class name can be
+        // extracted; or an allocation site of an object (i.e. a stack/global/heap variable), from which a
+        // forward walk can be performed to find calls to C++ constructor/destructor/template functions from
+        // which the class' name can then be extracted; skip starting pointer
         if (val == thisPtr) continue;
 
         if (const auto *func = SVFUtil::dyn_cast<Function>(val))
         {
-            // extract class name from function name
-            Set<std::string> classNames = extractClsNamesFromFunc(func);
-            insertClassNames(classNames);
+            // Constructor/destructor/template func; extract name from func directly
+            addNamesFromFunc(func);
         }
-        else if (SVFUtil::isa<LoadInst, StoreInst, GetElementPtrInst, AllocaInst, GlobalValue>(val))
+        else if (isClsNameSource(val))
         {
-            // extract class name from instructions
-            const Type *type = infersiteToType(val);
-            const std::string &className = typeToClsName(type);
```

Comment:
typeToClsName can be deleted if there are no uses.

Reply:
Removed `typeToClsName`; `infersiteToType()` still has remaining uses.